src/Build/BuildCheck/OM/BuildCheckDataContext.cs

```diff
@@ -19,10 +20,18 @@ namespace Microsoft.Build.Experimental.BuildCheck;
 /// <param name="projectFilePath">Currently built project.</param>
 public abstract class AnalysisData(string projectFilePath)
 {
+    private string? _projectFileDirectory;
+
     /// <summary>
     /// Full path to the project file being built.
     /// </summary>
     public string ProjectFilePath { get; } = projectFilePath;
+
+    /// <summary>
+    /// Directory path of the file being built (the containing directory of <see cref="ProjectFilePath"/>).
+    /// </summary>
+    public string ProjectFileDirectory =>
+        _projectFileDirectory ??= Path.GetDirectoryName(ProjectFilePath)!;
```

Comment:
If we really wanted to get fancy and avoid most of these allocations, we could run `GetDirectoryName` only once and reuse for all instances but I think this is good enough. Too bad that `Path.GetFullPath(..., Path.GetDirectoryName(...))` cannot be done with Span.

Reply:
this is great as is